File: app/src/main/java/com/pr0gramm/app/ui/fragments/LoaderHelper.java
package com.pr0gramm.app.ui.fragments;

import com.google.common.base.Optional;
import com.google.common.base.Supplier;

import rx.Observable;
import rx.Subscription;
import rx.android.schedulers.AndroidSchedulers;
import rx.functions.Action0;
import rx.functions.Action1;

/**
 */
public abstract class LoaderHelper<T> {
    private Optional<T> cachedValue;
    private Throwable cachedError;

    private Action1<T> valueCallback;
    private Action1<Throwable> errorCallback;

    private Subscription subscription;

    abstract protected Observable<T> newObservable();

    public void load(Action1<T> valueCallback, Action1<Throwable> errorCallback) {
        register(valueCallback, errorCallback, true);
    }

    public void load(Action1<T> valueCallback, Action1<Throwable> errorCallback,
                     Action0 completeCallback) {

        register(doFinally(valueCallback, completeCallback),
                doFinally(errorCallback, completeCallback),
                true);
    }

    public void register(Action1<T> valueCallback, Action1<Throwable> errorCallback) {
        register(valueCallback, errorCallback, false);
    }

    public void register(Action1<T> valueCallback, Action1<Throwable> errorCallback,
                         Action0 completeCallback) {

        register(doFinally(valueCallback, completeCallback),
                doFinally(errorCallback, completeCallback),
                false);
    }

    private void register(Action1<T> valueCallback, Action1<Throwable> errorCallback,
                          boolean startLoading) {

        // store for later
        this.valueCallback = valueCallback;
        this.errorCallback = errorCallback;

        if (cachedValue != null) {
            valueCallback.call(cachedValue.orNull());

        } else if (cachedError != null) {
            errorCallback.call(cachedError);

        } else if (subscription == null && startLoading) {
            reload();
        }
    }

    public void reload() {
        if (subscription != null) {
            subscription.unsubscribe();
        }

        subscription = newObservable()
                .first()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::onNext, this::onError);
    }

    public void detach() {
        valueCallback = null;
        errorCallback = null;
    }

    private void onNext(T value) {
        subscription = null;

        cachedValue = Optional.fromNullable(value);
        cachedError = null;

        if (valueCallback != null) {
            valueCallback.call(value);
        }
    }

    private void onError(Throwable error) {
        subscription = null;

        cachedError = error;
        cachedValue = null;

        if (errorCallback != null) {
            errorCallback.call(error);
        }
    }

    public static <T> LoaderHelper<T> of(Supplier<Observable<T>> supplier) {
        return new LoaderHelper<T>() {
            @Override
            protected Observable<T> newObservable() {
                return supplier.get();
            }
        };
    }

    public static <T> LoaderHelper<T> of(Observable<T> staticObservable) {
        Observable<T> cached = staticObservable.cache();
        return of(() -> cached);
    }

    private static <T> Action1<T> doFinally(Action1<T> firstAction, Action0 finallyAction) {
        return value -> {
            try {
                firstAction.call(value);
            } finally {
                finallyAction.call();
            }
        };
    }
}

File: app/src/main/java/com/pr0gramm/app/ui/fragments/LoaderHelper.kt
package com.pr0gramm.app.ui.fragments

import com.pr0gramm.app.util.BackgroundScheduler
import com.pr0gramm.app.util.MainThreadScheduler
import rx.Observable
import rx.Subscription
import rx.functions.Action0
import rx.functions.Action1

/**
 */
abstract class LoaderHelper<T> {
    private var cachedValue: T? = null
    private var cachedError: Throwable? = null

    private var valueCallback: Action1<T>? = null
    private var errorCallback: Action1<Throwable>? = null

    private var subscription: Subscription? = null

    protected abstract fun newObservable(): Observable<T>

    fun load(valueCallback: Action1<T>, errorCallback: Action1<Throwable>) {
        register(valueCallback, errorCallback, true)
    }

    fun load(valueCallback: Action1<T>, errorCallback: Action1<Throwable>,
             completeCallback: Action0) {

        register(doFinally(valueCallback, completeCallback),
                doFinally(errorCallback, completeCallback),
                true)
    }

    fun register(valueCallback: Action1<T>, errorCallback: Action1<Throwable>) {
        register(valueCallback, errorCallback, false)
    }

    fun register(valueCallback: Action1<T>, errorCallback: Action1<Throwable>,
                 completeCallback: Action0) {

        register(doFinally(valueCallback, completeCallback),
                doFinally(errorCallback, completeCallback),
                false)
    }

    private fun register(valueCallback: Action1<T>, errorCallback: Action1<Throwable>,
                         startLoading: Boolean) {

        // store for later
        this.valueCallback = valueCallback
        this.errorCallback = errorCallback

        if (cachedValue != null) {
            valueCallback.call(cachedValue)

        } else if (cachedError != null) {
            errorCallback.call(cachedError)

        } else if (subscription == null && startLoading) {
            reload()
        }
    }

    fun reload() {
        if (subscription != null) {
            subscription!!.unsubscribe()
        }

        subscription = newObservable()
                .take(1)
                .subscribeOn(BackgroundScheduler)
                .unsubscribeOn(BackgroundScheduler)
                .observeOn(MainThreadScheduler)
                .subscribe({ value -> this@LoaderHelper.onNext(value) }, { error -> this@LoaderHelper.onError(error) })
    }

    fun detach() {
        valueCallback = null
        errorCallback = null
    }

    private fun onNext(value: T) {
        subscription = null

        cachedValue = value
        cachedError = null

        if (valueCallback != null) {
            valueCallback!!.call(value)
        }
    }

    private fun onError(error: Throwable) {
        subscription = null

        cachedError = error
        cachedValue = null

        if (errorCallback != null) {
            errorCallback!!.call(error)
        }
    }

    companion object {

        fun <T> of(supplier: () -> Observable<T>): LoaderHelper<T> {
            return object : LoaderHelper<T>() {
                override fun newObservable(): Observable<T> {
                    return supplier()
                }
            }
        }

        private fun <T> doFinally(firstAction: Action1<T>, finallyAction: Action0): Action1<T> {
            return Action1 { value ->
                try {
                    firstAction.call(value)
                } finally {
                    finallyAction.call()
                }
            }
        }
    }
}
